datafusion/src/row/mod.rs

```diff
@@ -0,0 +1,334 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! An implementation of Row backed by raw bytes
+//!
+//! Each tuple consists of up to three parts: [null bit set] [values] [var length data]
+//!
+//! The null bit set is used for null tracking and is aligned to 1-byte. It stores
+//! one bit per field.
+//!
+//! In the region of the values, we store the fields in the order they are defined in the schema.
```

Comment:
This describes a variable length tuple -- so if you pack a bunch of `Row`s together in memory it will not be possible to find some arbitrary location quickly

For example, finding where `Row3 `starts in the following picture needs to scan  all columns of Row1 and Row2

```
┌─────────────────────────────────┐
│              Row1               │
├──────────┬──────────────────────┤
│   Row2   │         Row3         │
├──────────┴─┬────────────────────┤
│    Row4    │        Row5        │
└────────────┴────────────────────┘
```


The benefit of this strategy is that tuple construction will be very fast and memory usage optimal

There are other strategies that have a fixed width tuples, as discussed on https://github.com/apache/arrow-datafusion/issues/1708 that have benefits (though are likely not as memory optimal)

I think getting optimal performance will come from being able to vectorize many operations for which fixed sized tuples are compelling:

```
┌───────────┬──┐               ┌─────────────┐
│     Row1  │  ├──────┐        │             │
├───────────┼──┤      │ ┌──────┼▶            │
│     Row2  │  │──────┼┐├──────┼─▶           │
├───────────┼──┤      │││      │             │
│     Row3  │  │──────┼┼┤      │  variable   │
├───────────┼──┤      └┼┼─────▶│ length area │
│     Row4  │  │───────┼┘      │             │
├───────────┼──┤       └──────▶│             │
│     Row5  │  │───────────┐   │             │
└───────────┴──┘           └───┼───────▶     │
                               └─────────────┘
```

Maybe I can find some time this weekend to play around with some ideas

Reply:
One classic database technique to pack such data into a single block (rather than a separate variable length area) is to preallocate the page (e.g. 32K or something) and then write rows into the front of the page, but filling the variable length area from the *back*.